### PR TITLE
Implement the artifact abstractor and resolver

### PR DESCRIPTION
**Pull request by ywk253100**
1. Define the interface for artifact abstractor and resolver
2. Implement the artifact abstractor
3. Implement the resolver for image with manifest v2

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

File: src/api/artifact/resolver/resolver.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+// Register resolver, one resolver can handle multiple media types for one kind of artifact
+func Register(resolver Resolver, mediaTypes ...string) error {
+	for _, mediaType := range mediaTypes {
+		_, exist := registry[mediaType]
+		if exist {
+			return fmt.Errorf("resolver to handle media type %s already exists", mediaType)
+		}
+		registry[mediaType] = resolver
+		log.Debugf("resolver to handle media type %s registered", mediaType)
```
Comment (renmaosheng) [inline src/api/artifact/resolver/resolver.go]: it is better to use info, so that the common workflow can be logged.

File: src/api/artifact/abstractor.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+func (a *abstractor) Abstract(ctx context.Context, repoFullName string, artifact *artifact.Artifact) error {
+	// read manifest content
+	manifestMediaType, content, err := a.blobReader.Read(repoFullName, artifact.Digest, true)
+	if err != nil {
+		return err
+	}
+	artifact.ManifestMediaType = manifestMediaType
+
+	switch artifact.ManifestMediaType {
+	// docker manifest v1
```
Comment (reasonerjt) [inline src/api/artifact/abstractor.go]: What if we do not support it?

File: src/api/artifact/blob/cache.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package blob
+
+// TODO add cache
```
Comment (reasonerjt) [inline src/api/artifact/blob/cache.go]: If we enable cache in registry, there's already a cache of manifest in redis.

File: src/api/artifact/blob/reader.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+		schema1.MediaTypeSignedManifest,
+		schema2.MediaTypeManifest,
+		v1.MediaTypeImageManifest,
+		manifestlist.MediaTypeManifestList,
+		v1.MediaTypeImageIndex,
+	}
+)
+
+// Reader read blob content
+type Reader interface {
```
Comment (reasonerjt) [inline src/api/artifact/blob/reader.go]: Conventionally when you call a type xxxReader it implements `io.Reader`

File: src/api/artifact/blob/reader.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	// Read the content of blob under the repository. Set the "manifest" as true if whether the blob is a manifest
+	Read(repoFullName, digest string, manifest bool) (mediaType string, content []byte, err error)
```
Comment (reasonerjt) [inline src/api/artifact/blob/reader.go]: Since it does not implement `io.Reader` we should add different funcs like `GetManifest` `GetConfig`...  it's easier for testing.

File: src/api/artifact/resolver/resolver.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+func init() {
+	register(image.NewManifestV2Resolver(), image.ManifestV2ResolverMediaTypes...)
+	register(image.NewManifestV1Resolver(), image.ManifestV1ResolverMediaTypes...)
+	register(image.NewIndexResolver(), image.IndexResolverMediaTypes...)
+}
+
+// Resolver resolves the detail information for a specific kind of artifact
+type Resolver interface {
+	// ArtifactType returns the type of artifact that the resolver handles
+	ArtifactType(context.Context) string
```
Comment (reasonerjt) [inline src/api/artifact/resolver/resolver.go]: Why is this needed?

File: src/api/artifact/resolver/resolver.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	// Resolve receives the manifest content, resolves the detail information
+	// from the manifest or the layers referenced by the manifest, and populates
+	// the detail information into the artifact
+	// The layers can be read by reader
+	// The "repoFullName" is the "project_name/repository_name" that the artifact belongs to
+	Resolve(ctx context.Context, manifest []byte, repoFullName string, artifact *artifact.Artifact, reader blob.Reader) error
+}
+
+func register(resolver Resolver, mediaTypes ...string) {
```
Comment (reasonerjt) [inline src/api/artifact/resolver/resolver.go]: calling a public func from a private one???

File: src/api/artifact/resolver/resolver.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	if err := Register(resolver, mediaTypes...); err != nil {
+		log.Errorf("failed to register resolver for artifact %s: %v", resolver.ArtifactType(nil), err)
+	}
+}
+
+// Register resolver, one resolver can handle multiple media types for one kind of artifact
+func Register(resolver Resolver, mediaTypes ...string) error {
```
Comment (reasonerjt) [inline src/api/artifact/resolver/resolver.go]: If it's public it has to be thread-safe.

File: src/api/artifact/blob/cache.go (hunk 3 above)
Comment (ywk253100) [reply src/api/artifact/blob/cache.go]: It's the cache for both manifest and config layer

File: src/api/artifact/resolver/resolver.go (hunk 6 above)
Comment (ywk253100) [reply src/api/artifact/resolver/resolver.go]: `ArtifactType` indicates what artifact type of the resolver handles. For `image.v1ManifestResolver`, `image.v2ManifestResolver` and `image.indexManifestResolver`, the artifact type are all `IMAGE`, the `abstractor` needs the information to populate artifact

File: src/api/artifact/resolver/resolver.go (hunk 7 above)
Comment (ywk253100) [reply src/api/artifact/resolver/resolver.go]: `register` is just a util function, it wraps the `Register`

File: src/api/artifact/abstractor.go (hunk 2 above)
Comment (ywk253100) [reply src/api/artifact/abstractor.go]: If we don't support it, the images with v1 manifest cannot be pushed into Harbor, and there may be some issue when upgrading if the images with v1 manifest already exist in Harbor 1.x.

File: src/api/artifact/resolver/resolver.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	for _, mediaType := range mediaTypes {
+		_, exist := registry[mediaType]
+		if exist {
+			return fmt.Errorf("resolver to handle media type %s already exists", mediaType)
```
Comment (wy65701436) [inline src/api/artifact/resolver/resolver.go]: why not just log and return nil here?

File: src/api/artifact/blob/reader.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+	// TODO read from cache first
+	client, err := coreutils.NewRepositoryClientForLocal("admin", repoFullName)
+	if err != nil {
+		return "", nil, err
+	}
+	if manifest {
+		_, mediaType, payload, err := client.PullManifest(digest, accept)
+		return mediaType, payload, err
+	}
+	_, reader, err := client.PullBlob(digest)
```
Comment (wy65701436) [inline src/api/artifact/blob/reader.go]: `reader` here is confusing with the struct name.

File: src/api/artifact/resolver/resolver.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+	// Resolve receives the manifest content, resolves the detail information
+	// from the manifest or the layers referenced by the manifest, and populates
+	// the detail information into the artifact
+	// The layers can be read by reader
+	// The "repoFullName" is the "project_name/repository_name" that the artifact belongs to
+	Resolve(ctx context.Context, manifest []byte, repoFullName string, artifact *artifact.Artifact, reader blob.Reader) error
+}
+
+func register(resolver Resolver, mediaTypes ...string) {
+	if err := Register(resolver, mediaTypes...); err != nil {
```
Comment (wy65701436) [inline src/api/artifact/resolver/resolver.go]:
```
if len(mediaTypes) == 0 {
  return nil
}
...
```

File: src/api/artifact/resolver/resolver.go [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+		log.Errorf("failed to register resolver for artifact %s: %v", resolver.ArtifactType(nil), err)
+	}
+}
+
+// Register resolver, one resolver can handle multiple media types for one kind of artifact
+func Register(resolver Resolver, mediaTypes ...string) error {
+	for _, mediaType := range mediaTypes {
```
Comment (wy65701436) [inline src/api/artifact/resolver/resolver.go]: if len(mediaTypes) == 0 {
  return nil
}
...

**Comment by heww** [comment]
Should these source codes move to pkg as `pkg/artifact/abstractor`?

File: src/api/artifact/resolver/resolver.go [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+	"context"
+	"fmt"
+	"github.com/goharbor/harbor/src/api/artifact/blob"
+	"github.com/goharbor/harbor/src/api/artifact/resolver/image"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/pkg/artifact"
+)
+
+var (
+	registry = map[string]Resolver{}
```
Comment (heww) [inline src/api/artifact/resolver/resolver.go]: rename `registry` to `resolvers`?

File: src/api/artifact/resolver/resolver.go (hunk 1 above)
Comment (ywk253100) [reply src/api/artifact/resolver/resolver.go]: Updated

File: src/api/artifact/blob/reader.go (hunk 4 above)
Comment (ywk253100) [reply src/api/artifact/blob/reader.go]: Renamed it to `Fetcher`

File: src/api/artifact/blob/reader.go (hunk 5 above)
Comment (ywk253100) [reply src/api/artifact/blob/reader.go]: Updated

File: src/api/artifact/resolver/resolver.go (hunk 8 above)
Comment (ywk253100) [reply src/api/artifact/resolver/resolver.go]: Done

File: src/api/artifact/blob/reader.go (hunk 10 above)
Comment (ywk253100) [reply src/api/artifact/blob/reader.go]: Updated

File: src/api/artifact/resolver/resolver.go (hunk 7 above)
Comment (heww) [reply src/api/artifact/resolver/resolver.go]: It should panic when buildin resolvers register failed. There is not any help to log error info when this problem happened.

File: src/api/artifact/resolver/resolver.go (hunk 6 above)
Comment (reasonerjt) [reply src/api/artifact/resolver/resolver.go]: So the values are static, for that case I don't think ctx is needed as a parm.

File: src/api/artifact/resolver/resolver.go (hunk 7 above)
Comment (reasonerjt) [reply src/api/artifact/resolver/resolver.go]: I was thinking we can make it private and expose it only when needed.  But that's trivial...

File: src/api/artifact/abstractor.go (hunk 2 above)
Comment (renmaosheng) [reply src/api/artifact/abstractor.go]: this is good question, I noticed that docker has deprecated the schemav1 support.

File: src/api/artifact/blob/cache.go (hunk 3 above)
Comment (steven-zou) [reply src/api/artifact/blob/cache.go]: Is it expensive for sending http request directly to get the data which is not large? If it is not, suggest removing cache out of the table so far. If one day, there is some perf issue appeared, then we can consider it at that time.

File: src/api/artifact/blob/fetcher.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+// Fetcher fetches the content of blob
+type Fetcher interface {
+	// FetchManifest the content of manifest under the repository
+	FetchManifest(repoFullName, digest string) (mediaType string, content []byte, err error)
+	// FetchLayer the content of layer under the repository
+	FetchLayer(repoFullName, digest string) (content []byte, err error)
+}
+
+// NewFetcher returns an instance of the default blob fetcher
+func NewFetcher() Fetcher {
```
Comment (steven-zou) [inline src/api/artifact/blob/fetcher.go]: If no cache enabled, fetcher is becoming unnecessary.

File: src/api/artifact/resolver/resolver.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+// Resolver resolves the detail information for a specific kind of artifact
+type Resolver interface {
+	// ArtifactType returns the type of artifact that the resolver handles
+	ArtifactType(context.Context) string
+	// Resolve receives the manifest content, resolves the detail information
+	// from the manifest or the layers referenced by the manifest, and populates
+	// the detail information into the artifact
+	// The layers can be fetched by fetcher
+	// The "repoFullName" is the "project_name/repository_name" that the artifact belongs to
+	Resolve(ctx context.Context, manifest []byte, repoFullName string, artifact *artifact.Artifact, fetcher blob.Fetcher) error
```
Comment (steven-zou) [inline src/api/artifact/resolver/resolver.go]: Based the function of `resovle` method as well as method meaning, I don't think so many parameters are required,

The following pattern is enough.
```go
Resolve(ctx context.Context, manifest []byte, artifact *artifact.Artifact) error
```

The `Fetcher` is just a util tool.

File: src/api/artifact/resolver/resolver.go (hunk 12 above)
Comment (steven-zou) [reply src/api/artifact/resolver/resolver.go]: we don't need it.

for loop will check it/

File: src/api/artifact/blob/cache.go (hunk 3 above)
Comment (ywk253100) [reply src/api/artifact/blob/cache.go]: As there is no OCI registry driver interface yet, we can keep the blob fetcher for now, and remove it util we define the driver interface

**Comment by ywk253100** [comment]
> Should these source codes move to pkg as `pkg/artifact/abstractor`?

No, the abstractor/resolver calls functions inside the pkg

File: src/api/artifact/resolver/resolver.go (hunk 6 above)
Comment (ywk253100) [reply src/api/artifact/resolver/resolver.go]: Updated
